store/src/main/java/org/apache/rocketmq/store/MappedFile.java

```diff
@@ -295,13 +295,15 @@ public int flush(final int flushLeastPages) {
 
                 try {
                     //We only append data to fileChannel or mappedByteBuffer, never both.
+                    this.flushError = false;
                     if (writeBuffer != null || this.fileChannel.position() != 0) {
                         this.fileChannel.force(false);
                     } else {
                         this.mappedByteBuffer.force();
                     }
                 } catch (Throwable e) {
                     log.error("Error occurred when force data to disk.", e);
+                    this.flushError = true;
```

Comment:
If this time flush fails, the next time flush succeeds, but every time you returns 'flush failed' to the client?

Reply:
@hzh0425   “this.flushError = false; ”   will  reset flushError   to  false  when  flush

Reply:
Flush error rarely happens. And the client does not need to know the flush error.
How about just making it simple, just do not change the flushed position when encountering some unexpected error.

Do not introduce a new response code, and do not need to change other files.

Reply:
@dongeforever Thank you for your review. What you said is also a good idea to solve the problem of data loss.very concise.
But IMO, if "just do not change the flushed position", it may cause "FLUSH_DISK_TIMEOUT" in every flush. when  FLUSH_DISK_TIMEOUT，even if the client resends, it may still be sent to the failed broker. Can it be optimized?

Reply:
@dongeforever When the storage fails, the delayed message cannot be delivered to the target topic, and the delayoffset.json will be synchronized to the slave normally, which will cause the slave to be unable to recover the delayed message .
In addition, the slave is also asynchronous flush. Relying on multiple nodes to ensure that data is not lost, kafka can also do it. The disadvantage of kafka is that there is no synchronous flush